src/main/scala/diplomaticobjectmodel/OMLogicalTree.scala

```diff
@@ -0,0 +1,72 @@
+// See LICENSE.SiFive for license details.
+
+package freechips.rocketchip.tile
+
+import freechips.rocketchip.diplomacy.ResourceBindingsMap
+import freechips.rocketchip.diplomaticobjectmodel.model.OMComponent
+import scala.collection.mutable.ArrayBuffer
+
+
+trait LogicalTree {
+  def getOMComponents(resourceBindingsMap: ResourceBindingsMap, children: Seq[OMComponent]): Seq[OMComponent]
+}
+
+/** LogicalTreeEdges hold LogicalTree Nodes which will be used to construct the logical tree.
+  * The LogicalTreeEdge is used to construct a parent child relationship between different modules.
+  *
+  * @param parent The parent LogicalTree node
+  * @param child The child LogicalTree node
+  */
+case class LogicalTreeEdge(
+  parent: LogicalTree,
+  child: LogicalTree
+)
+
+object LogicalModuleTree {
+  val edges = ArrayBuffer[LogicalTreeEdge]()
+
+  def add(parent: LogicalTree, child: LogicalTree): Unit = {
+     edges += LogicalTreeEdge(parent, child)
+  }
+
+  /**
+    * The getTreeMap function constructs a tree by converting the list of LogicalTreeEdges into a Map which
+    * contains each of the parents and their children
+    *
+    * @return
+    */
+  def getTreeMap(): Map[LogicalTree, List[LogicalTree]] = {
+    edges.groupBy(_.parent).map{ case (k, v) => (k, v.map(_.child).toList)}
+  }
+
+  def findRoot(): LogicalTree = {
+    val values = getTreeMap().values.flatten.toSet
+    val keys = getTreeMap().keys.toSet
+
+    val roots = values.diff(keys)
+
+    assert(roots.size == 1, "Logical Tree contains more than one root.")
+    roots.head
+  }
+}
+
+case class Tree[A](parent: A, children: List[Tree[A]])
+
+object OMLogicalTreeTree {
+  def makeTree(): Tree[LogicalTree] = {
+    val root: LogicalTree = LogicalModuleTree.findRoot()
+    val treeMap = LogicalModuleTree.getTreeMap()
+
+    def recurse(treeNode: LogicalTree): Tree[LogicalTree] = {
```

Comment:
Shouldn't this be a method of `LogicalTree`? What is it used for?

Reply:
See the above comment.